notebook/services/kernels/kernelmanager.py

```diff
@@ -156,38 +304,49 @@ def start_kernel(self, kernel_id=None, path=None, **kwargs):
             an existing kernel is returned, but it may be checked in the future.
         """
         if kernel_id is None:
-            if path is not None:
-                kwargs['cwd'] = self.cwd_for_path(path)
-            kernel_id = yield gen.maybe_future(
-                super(MappingKernelManager, self).start_kernel(**kwargs)
-            )
-            self._kernel_connections[kernel_id] = 0
-            self.start_watching_activity(kernel_id)
-            self.log.info("Kernel started: %s" % kernel_id)
-            self.log.debug("Kernel args: %r" % kwargs)
-            # register callback for failed auto-restart
-            self.add_restart_callback(kernel_id,
-                lambda : self._handle_kernel_died(kernel_id),
-                'dead',
-            )
-
-            # Increase the metric of number of kernels running
-            # for the relevant kernel type by 1
-            KERNEL_CURRENTLY_RUNNING_TOTAL.labels(
-                type=self._kernels[kernel_id].kernel_name
-            ).inc()
-
+            kernel_id = self.start_launching_kernel(path, kernel_name, **kwargs)
+            yield self.get_kernel(kernel_id).client_ready()
         else:
             self._check_kernel_id(kernel_id)
             self.log.info("Using existing kernel: %s" % kernel_id)
 
-        # Initialize culling if not already
-        if not self._initialized_culler:
-            self.initialize_culler()
-
         # py2-compat
         raise gen.Return(kernel_id)
 
+    def start_launching_kernel(self, path=None, kernel_name=None, **kwargs):
+        """Launch a new kernel, return its kernel ID
+
+        This is a synchronous method which starts the process of launching a
+        kernel. Retrieve the KernelInterface object and call ``.client_ready()``
+        to get a future for the rest of the startup & connection.
+        """
+        if path is not None:
+            kwargs['cwd'] = self.cwd_for_path(path)
+        kernel_id = str(uuid.uuid4())
```

Comment:
Please restore the call to [new_kernel_id()](https://github.com/jupyter/jupyter_client/blob/master/jupyter_client/multikernelmanager.py#L319-L326).  It's important that applications be able to specify their own kernel_id.  Another option may be to let the kernel provider determine its kernel id - assuming the provider also has access to `kwargs` (see next comment).

Reply:
That should be easy enough. Can you point me to the code where you're overriding it, out of interest?

Reply:
Sure: https://github.com/jupyter/enterprise_gateway/blob/master/enterprise_gateway/services/kernels/remotemanager.py#L93

Reply:
That code confuses me a bit. Normally the point of a `MappingKernelManager` is to manage multiple kernels with different IDs, but if you're pulling the kernel ID from an environment variable, it will always be the same (unless it's being changed in process, which hopefully isn't the case). So you can only ever have one kernel in the manager? So why do you need the MappingKernelManager?

Reply:
(Sorry for the delayed response)
The `MappingKernelManager` (well, actually, the previous `MultiKernelManager`) is responsible for equating a kernel manager instance with a kernel_id.  

The environment for which this value is derived is a portion of the kernel creation request that comes from clients of Enterprise Gateway.  In these applications, clients can influence the request.  It is not the environment of the process itself.  

As was noted in the next comment, by specifying things like kernel_id and kernel namespace, for example, clients can "seed" environments in which their kernels will be invoked.  Since kernel_id is a perfect "key", allowing clients to specify their kernel_id is critical to enabling them to build applications as they desire based on that value.

Please restore that functionality as it is current behavior.